cglib/src/main/java/net/sf/cglib/proxy/Enhancer.java

```diff
@@ -413,10 +394,58 @@ private void validate() {
 
     static class EnhancerFactoryData {
         public final Class generatedClass;
-        public volatile Factory factory;
+        private final Method setThreadCallbacks;
+        private final Class[] primaryConstructorArgTypes;
+        private final Constructor primaryConstructor;
 
-        public EnhancerFactoryData(Class generatedClass) {
+        public EnhancerFactoryData(Class generatedClass, Class[] primaryConstructorArgTypes, boolean classOnly) {
             this.generatedClass = generatedClass;
+            try {
+                setThreadCallbacks = getCallbacksSetter(generatedClass, SET_THREAD_CALLBACKS_NAME);
+                setThreadCallbacks.setAccessible(true);
+                if (classOnly) {
+                    this.primaryConstructorArgTypes = null;
+                    this.primaryConstructor = null;
+                } else {
+                    this.primaryConstructorArgTypes = primaryConstructorArgTypes;
+                    this.primaryConstructor = generatedClass.getDeclaredConstructor(primaryConstructorArgTypes);
+                    primaryConstructor.setAccessible(true);
+                }
+            } catch (NoSuchMethodException e) {
+                throw new CodeGenerationException(e);
+            }
+        }
+
+        public Object newInstance(Class[] argumentTypes, Object[] arguments, Callback[] callbacks) {
+            setThreadCallbacks(callbacks);
+            try {
+                if (!Arrays.equals(primaryConstructorArgTypes, argumentTypes)) {
+                    return ReflectUtils.newInstance(generatedClass, argumentTypes, arguments);
+                }
+                try {
+                    return primaryConstructor.newInstance(arguments);
```

Comment:
Technically speaking, it tries to revert setAccessible as well. I'm not sure what implications that would have in multithreaded environment where several threads would want to alter accessibility of the same instance at exactly the same time.